kong/db/declarative/init.lua

```diff
@@ -908,12 +904,46 @@ do
              "incomplete default workspace id sent to the stream subsystem")
     end
 
+    return true
+  end
+
+  -- If it takes more than 60s it is very likely to be an internal error.
+  -- However it will be reported as: "failed to broadcast reconfigure event: recursive".
+  -- Let's paste the error message here in case someday we try to search it.
+  -- Should we handle this case specially?
+  local DECLARATIVE_LOCK_TTL = 60
+  local DECLARATIVE_MAX_RETRT = 10
+  local DECLARATIVE_LOCK_KEY = "declarative:lock"
 
+  -- make sure no matter which path it exits, we released the lock.
+  function declarative.load_into_cache_with_events(entities, meta, hash)
     if exiting() then
       return nil, "exiting"
     end
 
-    return true
+    local kong_shm = ngx.shared.kong
+
+    local ok, err = kong_shm:add(DECLARATIVE_LOCK_KEY, 0, DECLARATIVE_LOCK_TTL)
+    if not ok then
+      if err == "exists" then
+        local ttl = math.min(ngx.shared.kong:ttl(DECLARATIVE_LOCK_KEY), DECLARATIVE_MAX_RETRT)
+        return nil, "busy", ttl
+      end
+
+      kong_shm:delete(DECLARATIVE_LOCK_KEY)
+      return nil, err
+    end
+
+    ok, err = load_into_cache_with_events_no_lock(entities, meta, hash)
+
+    kong_shm:delete(DECLARATIVE_LOCK_KEY)
+
+
+    if exiting() then
+      return nil, "exiting"
+    end
```

Comment:
We should move this back to the beginning of `load_into_cache_with_events_no_lock`, before the config is actually loaded.

Reply:
I've moved two of them back. Though I don't know why we need to check again after the config loaded, this is what it was like before this patch.

Reply:
I've moved two of them back. Though I don't know why we need to check again after the config loaded, this is what it was like before this patch.